### PR TITLE
Implement if flattening to address if-blocks that may have mutation

### PR DESCRIPTION
**Pull request by swapnilraj**
Flatten if-blocks such that they are abstracted away to functions with only one if statement

### CODE REVIEW

File: warp/yul/ScopeFlattener.py [hunk 1]
```diff
@@ -47,13 +47,78 @@ def visit_block(self, node: ast.Block, inline: Optional[bool] = None):
...
             return_variables=self.visit_list(node.return_variables),
             body=self.visit(node.body, inline=True),
         )
 
+    def visit_if(self, node: ast.If):
+        free_vars = set()
+        mod_vars = set()
+
+        free_vars.update(node.body.scope.free_variables)
+        free_vars.update(node.body.scope.modified_variables)
```
Comment (murcake) [inline warp/yul/ScopeFlattener.py:87]: I think it's unnecessary, modified_variables are always a subset of free_variables (to be more precise, these are free variables that has been modified in the scope). I should have written it in the invariant of Scope.

File: warp/yul/ScopeFlattener.py [hunk 2]
```diff
@@ -47,13 +47,78 @@ def visit_block(self, node: ast.Block, inline: Optional[bool] = None):
...
+        mod_vars.update(node.body.scope.modified_variables)
+
+        free_vars.update(ast.Block((node.condition,)).scope.free_variables)
+
+        if node.else_body:
+            free_vars.update(node.else_body.scope.free_variables)
+            free_vars.update(node.else_body.scope.modified_variables)
+            mod_vars.update(node.else_body.scope.modified_variables)
```
Comment (murcake) [inline warp/yul/ScopeFlattener.py:95]: A curious thought: you can get all free and modified variables in a more terse way by simply wrapping the whole `node` into a `Block`. Then you call `.scope` and here you are

**Review by murcake — APPROVED** [review]

File: warp/yul/ScopeFlattener.py (hunk 2 above)
Comment (swapnilraj) [reply warp/yul/ScopeFlattener.py:95]: Yea that seems like a better way; I'll add those changes

File: warp/yul/ScopeFlattener.py (hunk 1 above)
Comment (swapnilraj) [reply warp/yul/ScopeFlattener.py:87]: Oh okay.

**Review by BitBaseBit — CHANGES_REQUESTED** [review]
Throws this error for the ERC20 example:
 ```
 python warp/yul/main.py examples/ERC20/ERC20.sol WARP > res.cairo
 starknet-compile res.cairo --cairo_path=warp/cairo-src
res.txt:405:9: Unknown identifier '__warp_block_10'.
        __warp_block_10(_1)
        ^*************^
 ```

**Comment by swapnilraj** [comment]
> Throws this error for the ERC20 example:
> 
> ```
> python warp/yul/main.py examples/ERC20/ERC20.sol WARP > res.cairo
> starknet-compile res.cairo --cairo_path=warp/cairo-src
> res.txt:405:9: Unknown identifier '__warp_block_10'.
>        __warp_block_10(_1)
>        ^*************^
> ```

The new commit fixes this.

**Comment by swapnilraj** [comment]
> Throws this error for the ERC20 example:
> 
> ```
> python warp/yul/main.py examples/ERC20/ERC20.sol WARP > res.cairo
> starknet-compile res.cairo --cairo_path=warp/cairo-src
> res.txt:405:9: Unknown identifier '__warp_block_10'.
>        __warp_block_10(_1)
>        ^*************^
> ```

**Pull request merged**